### PR TITLE
functional courtyards calculation

### PR DESCRIPTION
**Pull request by u3ks**
Refactoring courtyards calculation to fit with the functional API and to speed it up.

### CODE REVIEW

File: momepy/functional/_intensity.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
+import geopandas as gpd
+from geopandas import GeoDataFrame, GeoSeries
+from libpysal.graph import Graph
+from packaging.version import Version
+from pandas import Series
+from shapely import buffer, get_num_interior_rings, unary_union
+
+__all__ = ["courtyards"]
+
+GPD_013 = Version(gpd.__version__) >= Version("0.13")
```
Comment (martinfleis) [inline momepy/functional/_intensity.py:10]:
```suggestion
```


Not used now.

File: momepy/functional/_intensity.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
+import geopandas as gpd
+from geopandas import GeoDataFrame, GeoSeries
+from libpysal.graph import Graph
+from packaging.version import Version
```
Comment (martinfleis) [inline momepy/functional/_intensity.py:4]:
```suggestion
```

File: momepy/functional/_intensity.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+    Parameters
+    ----------
+    geometry : GeoDataFrame
+        A GeoDataFrame containing objects to analyse.
+    graph : libpysal.graph.Graph
+        A spatial weights matrix for the geodataframe,
+        it is used to denote adjacent buildings.
+    Returns
+    ----------
```
Comment (martinfleis) [inline momepy/functional/_intensity.py:26]:
```suggestion

    Returns
    -------
```

File: momepy/functional/_intensity.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    series : Series
```
Comment (martinfleis) [inline momepy/functional/_intensity.py:27]:
```suggestion
    Series
```

File: momepy/functional/_intensity.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        A Series containing the resulting values.
+
+    Examples
+    --------
+    >>> courtyards = mm.calculate_courtyards(buildings_df, graph)"""
```
Comment (martinfleis) [inline momepy/functional/_intensity.py:32]:
```suggestion
    >>> courtyards = mm.calculate_courtyards(buildings_df, graph)
    """
```

We will eventually want this executable but that can be done together with other examples.

File: momepy/functional/_intensity.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+    # helper function to carry out the per group calculations
+    def _calculate_courtyards(group):
+        return get_num_interior_rings(unary_union(buffer(group.values, 0.01)))
```
Comment (martinfleis) [inline momepy/functional/_intensity.py:36]:
```suggestion
        return shapely.get_num_interior_rings(shapely.union_all(shapely.buffer(group.values, 0.01)))
```

**Comment by u3ks** [comment]
Timings:

new time: 11.6 s ± 191 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
old time: 43.8 s ± 635 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

**Review by jGaboardi — APPROVED** [review]

**Review by martinfleis — APPROVED** [review]

**Pull request merged**